Gym/DQN/main.swift

```diff
@@ -0,0 +1,278 @@
+            let leftQValue = Float(qValues[0][0]).unwrapped()
+            let rightQValue = Float(qValues[0][1]).unwrapped()
+            return leftQValue < rightQValue ? Tensor<Int32>(numpy: np.array(1, dtype: np.int32))! : Tensor<Int32>(numpy: np.array(0, dtype: np.int32))!
+        }
+    }
+
+    func train(batchSize: Int) {
+        // Don't train if replay buffer is too small
+        if replayBuffer.count >= batchSize {
+            // print("train | Start training")
+            let (tfStateBatch, tfActionBatch, tfRewardBatch, tfNextStateBatch) = replayBuffer.sample(batchSize: batchSize)
+
+            // TODO: Find equivalent function of tf.gather_nd in S4TF to parallelize Q-value computation (_Raw.gather_nd does not exist)
+            // Gradient are accumulated since we calculate every element in the batch individually
+            var totalGrad = qNet.zeroTangentVector
+            for i in 0..<batchSize {
+                let 𝛁qNet = gradient(at: qNet) { qNet -> Tensor<Float> in
+
+                    let stateQValueBatch = qNet(tfStateBatch)
+                    let tfAction: Tensor<Int32> = tfActionBatch[i][0]
+                    let action = Int(tfAction.makeNumpyArray()).unwrapped()
+                    let prediction: Tensor<Float> = stateQValueBatch[i][action]
+
+                    let nextStateQValueBatch = self.targetQNet(tfNextStateBatch)
+                    let tfReward: Tensor<Float> = tfRewardBatch[i][0]
+                    let leftQValue = Float(nextStateQValueBatch[i][0].makeNumpyArray()).unwrapped()
+                    let rightQValue = Float(nextStateQValueBatch[i][1].makeNumpyArray()).unwrapped()
+                    let maxNextStateQValue = leftQValue > rightQValue ? leftQValue : rightQValue
+                    let target: Tensor<Float> = tfReward + self.discount * maxNextStateQValue
+
+                    return squaredDifference(prediction, withoutDerivative(at: target))
+                }
+                totalGrad += 𝛁qNet
+            }
+            optimizer.update(&qNet, along: totalGrad)
+        }
+    }
+}
+
+func updateTargetQNet(source: Net, target: inout Net) {
+    target.l1.weight = Tensor<Float>(source.l1.weight)
+    target.l1.bias = Tensor<Float>(source.l1.bias)
+    target.l2.weight = Tensor<Float>(source.l2.weight)
+    target.l2.bias = Tensor<Float>(source.l2.bias)
+}
+
+class TensorFlowEnvironmentWrapper {
+    let originalEnv: PythonObject
+    let action_space: PythonObject
+    let observation_space: PythonObject
+
+    init(_ env: PythonObject) {
+        self.originalEnv = env
+        self.action_space = env.action_space
+        self.observation_space = env.observation_space
+    }
+
+    func reset() -> Tensor<Float> {
+      let state = self.originalEnv.reset()
+      return Tensor<Float>(numpy: np.array(state, dtype: np.float32))!
+    }
+
+    func step(_ action: Tensor<Int32>) -> (Tensor<Float>, Tensor<Float>, PythonObject, PythonObject) {
```

Comment:
Please add labels to the tuple result type! From the [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines/#label-closure-parameters):

> **Label tuple members and name closure parameters** where they appear in your API.
>
> These names have explanatory power, can be referenced from documentation comments, and provide expressive access to tuple members.

Since `TensorFlowEnvironmentWrapper` is a Swift wrapper around Python objects, I also think it makes sense to change the type of the "isDone" tuple element from `PythonObject` to `Bool`.

```suggestion
    func step(_ action: Tensor<Int32>) -> (state: Tensor<Float>, reward: Tensor<Float>, isDone: Bool, info: PythonObject) {
```

Reply:
Didn't know that I could add labels! Will do: they definitely make it more readable.

I realized that I did not used `isDone` in the target calculation. The correct target calculation is

reward + isDone * discount * max_a Q(s', a)

So `isDone` should probably be a `Tensor<Bool>` or `Tensor<Float>`. Thanks for pointing it out!